### ISSUE TITLE
trial -e doesn't work with pyunit test cases

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1837]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)| @radix reported|
|-|-|
|Trac ID|trac#1837|
|Type|defect|
|Created|2006-06-21 17:41:08Z|

More frame-trimming stuff!

```
radix@naota ~% cat test_crap.py
import unittest
class Foo(unittest.TestCase):
    def test_crap(self):
        1/0

radix@naota ~% trial -e test_crap.py
Running 1 tests.
test_crap
  Foo
    test_crap ... Traceback (most recent call last):
  File "/home/radix/Projects/Twisted/trunk/bin/trial", line 24, in ?
    run()
  File "/home/radix/Projects/Twisted/trunk/twisted/scripts/trial.py", line 377, in run
    test_result = trialRunner.run(suite)
  File "/home/radix/Projects/Twisted/trunk/twisted/trial/runner.py", line 499, in run
    suite.run(result)
  File "/home/radix/Projects/Twisted/trunk/twisted/trial/runner.py", line 179, in run
    TestSuite.run(self, result)
  File "/home/radix/Projects/Twisted/trunk/twisted/trial/runner.py", line 117, in run
    test(result)
  File "/home/radix/Projects/Twisted/trunk/twisted/trial/runner.py", line 109, in __call__
    return self.run(result)
  File "/home/radix/Projects/Twisted/trunk/twisted/trial/runner.py", line 117, in run
    test(result)
  File "/home/radix/Projects/Twisted/trunk/twisted/trial/runner.py", line 109, in __call__
    return self.run(result)
  File "/home/radix/Projects/Twisted/trunk/twisted/trial/runner.py", line 117, in run
    test(result)
  File "/home/radix/Projects/Twisted/trunk/twisted/trial/runner.py", line 109, in __call__
    return self.run(result)
  File "/home/radix/Projects/Twisted/trunk/twisted/trial/runner.py", line 117, in run
    test(result)
  File "/usr/lib/python2.4/unittest.py", line 281, in __call__
    return self.run(*args, **kwds)
  File "/usr/lib/python2.4/unittest.py", line 267, in run
    result.addError(self, self.__exc_info())
  File "/home/radix/Projects/Twisted/trunk/twisted/trial/reporter.py", line 434, in addError
    super(TreeReporter, self).addError(*args)
  File "/home/radix/Projects/Twisted/trunk/twisted/trial/reporter.py", line 171, in addError
    self.write(self._formatFailureTraceback(error))
  File "/home/radix/Projects/Twisted/trunk/twisted/trial/reporter.py", line 245, in _formatFailureTraceback
    fail.frames, frames = self._trimFrames(fail.frames), fail.frames
AttributeError: 'tuple' object has no attribute 'frames'
radix@naota ~%

```

<details><summary>Searchable metadata</summary>

```
trac-id__1837 1837
type__defect defect
reporter__radix radix
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1150911668000000 1150911668000000
changetime__1154149256000000 1154149256000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
Index: twisted/trial/reporter.py
===================================================================
--- twisted/trial/reporter.py   (revision 17324)
+++ twisted/trial/reporter.py   (working copy)
@@ -168,7 +168,10 @@
     def addError(self, test, error):
         super(Reporter, self).addError(test, error)
         if self.realtime:
-            self.write(self._formatFailureTraceback(error))
+            # use the error that addError appended instead of the one
+            # that was passed to us in case it was converted from a
+            # tuple to a failure
+            self.write(self._formatFailureTraceback(self.errors[-1][1]))

     def write(self, format, *args):
         s = str(format)
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @radix|
|-|-|

REJECT!  No unit test.

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

you are wise!

It doesn't handle the failure case, either.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @radix|
|-|-|

Branch in trial-pyunit-realtime-1837

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix set owner to @jml|
|-|-|

I've tried this on my codebase which uses trial and it works great. The added test looks kind of strange, given that it does nothing explicit about testing the real-timeness, but I guess it a test that errored before the change and now passes.  If you can think of a way to test it explicitly then that'd be cool but I think it's mergeable now.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set status to `closed`|
|-|-|

(In [17727]) Allow errors from pyunit tests to be reported realtime.

 * Author: jml
 * Reviewer: radix
 * Fixes [#1837](https://github.com/twisted/twisted/issues/1837)

Trial would blow up when reporting failures from pyunit tests ''only
in realtime''.  This branch wraps the pyunit exception info in a failure
object, so the stack trimming goodness works properly.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
